Session generator - NoMethodError on sign out

yshmarov opened the pull request:
When an authenticated user clicks `button_to 'Sign out', session_path, method: :delete>, he gets an error:

<img width="696" alt="Screenshot 2024-07-19 at 21 39 54" src="https://github.com/user-attachments/assets/7dcdd078-1767-4468-99bd-31b12642b0e3">

This will fix it.

yshmarov committed:
session generator - fix sign out

undefined method 'destroy' for nil:NilClass

yshmarov committed:
request_authentication - add alert

notify user that something was wrong, not just redirect

yshmarov committed:
do not define unused routes

dixpac commented on railties/lib/rails/generators/rails/sessions/templates/controllers/sessions_controller.rb:
```diff
@@ -1,5 +1,5 @@
 class SessionsController < ApplicationController
-  allow_unauthenticated_access
+  allow_unauthenticated_access(only: %i[new create])
```
👍

It is a style preference but to me this looks cleaner:
`allow_unauthenticated_access only: %i[ new create ]`

yshmarov replied:
Agree. Updated. Now it will be consistent with [this](https://github.com/rails/rails/blob/main/railties/lib/rails/generators/rails/scaffold_controller/templates/controller.rb.tt#L3)

dixpac commented on railties/lib/rails/generators/rails/sessions/templates/controllers/concerns/authentication.rb:
```diff
@@ -37,7 +37,7 @@ def find_session_by_cookie
     def request_authentication
       session[:return_to_after_authenticating] = request.url
-      redirect_to new_session_url
+      redirect_to new_session_url, alert: "Please log in to continue."
```
Maybe a bit too much with the alert here. When app redirects to sign in screen it is clear what happened without the **alert**

yshmarov replied:
I think it's a good pattern to add flash messages to redirects like "require authentication/require authentication". Anyway, it's quite out of scope of what I initially wanted to fix with this PR -> removed 🙌

dixpac reviewed (commented):
Good catch

yshmarov committed:
lint sessions controller

yshmarov committed:
authenticated users can not access sessions#new

authenticated? was always nil, because `allow_unauthenticated_access` skips `resume_session`

yshmarov committed:
remove extra alert

out of scope

yshmarov commented on railties/lib/rails/generators/rails/sessions/templates/controllers/sessions_controller.rb:
```diff
@@ -1,5 +1,6 @@
 class SessionsController < ApplicationController
-  allow_unauthenticated_access
+  allow_unauthenticated_access(only: %i[ new create ])
+  before_action :resume_session, only: :new
```
without this line, `authenticate?` in the `#new` action will always return `nil`

dixpac replied:
Yup!
I've submitted similar PR(https://github.com/rails/rails/pull/52373) for this yesterday, I'm ok with everything being contained in this PR tho, I can close my PR

brunoprietog replied:
I'd solve this other problem at a more general level, within the same concern

skipkayhil replied:
The `resume_session` bit was fixed in 1dd37e01937f6711cd15267481df8c0446961472. Can you rebase and update the PR title/description with a summary of the remaining changes?

brunoprietog commented on railties/lib/rails/generators/rails/sessions/templates/controllers/sessions_controller.rb:
```diff
@@ -1,5 +1,6 @@
 class SessionsController < ApplicationController
-  allow_unauthenticated_access
+  allow_unauthenticated_access(only: %i[ new create ])
```
```suggestion
  allow_unauthenticated_access only: %i[ new create ]
```

dhh commented on railties/lib/rails/generators/rails/sessions/sessions_generator.rb:
```diff
@@ -16,7 +16,7 @@ def create_session_files
       def configure_application
         gsub_file "app/controllers/application_controller.rb", /(class ApplicationController < ActionController::Base)/, "\\1\n  include Authentication"
-        route "resource :session"
+        route "resource :session, only: %i[new create destroy]"
```
Not interested in this change. No need to narrow this down. It pollutes the routes.rb file and the benefits are dubious.

yshmarov replied:
Since Rails [added](https://github.com/rails/rails/pull/45701) the `bin/rails routes --unused` command, I thought it makes sense to limit the unused routes

dhh replied:
Yeah, I think adding that might have been a mistake, if it drives people to overscope in their routes.rb. Meant to find routes for resources that don't exist at all any more.

dhh commented:
Good catch. Needs a rebase and the routes.rb change removed.
